sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala

```diff
@@ -317,22 +345,21 @@ case class AdaptiveSparkPlanExec(
    */
   private def updateLogicalPlan(
       logicalPlan: LogicalPlan,
-      newStages: Seq[(Exchange, QueryStageExec)]): LogicalPlan = {
+      newStages: Seq[QueryStageExec]): LogicalPlan = {
     var currentLogicalPlan = logicalPlan
     newStages.foreach {
-      case (exchange, stage) =>
-        // Get the corresponding logical node for `exchange`. If `exchange` has been transformed
-        // from a `Repartition`, it should have `logicalLink` available by itself; otherwise
-        // traverse down to find the first node that is not generated by `EnsureRequirements`.
-        val logicalNodeOpt = exchange.logicalLink.orElse(exchange.collectFirst {
-          case p if p.logicalLink.isDefined => p.logicalLink.get
-        })
+      case stage if currentPhysicalPlan.find(_.eq(stage)).isDefined =>
```

Comment:
will `currentPhysicalPlan.find(_.eq(stage))` be empty?

Reply:
Following this comment https://github.com/apache/spark/pull/25456/files#r314805998:

Now that we might have `newStages` from different rounds of stage creation, some stages might have been included in newer stages already, so those are not "reachable" now and we don't need to worry about them any more. But meanwhile we need to make sure we always apply the latest stages first: https://github.com/apache/spark/pull/25456/files#diff-6954dd8020a9ca298f1fb9602c0e831cR142.